luarules/gadgets/unit_xp_range_bonus.lua

```diff
@@ -14,20 +14,26 @@ if not gadgetHandler:IsSyncedCode() then
     return false
 end
 
-
 local SetUnitWeaponState = Spring.SetUnitWeaponState
 local GetUnitExperience = Spring.GetUnitExperience
 local SetUnitMaxRange = Spring.SetUnitMaxRange
 local GetUnitDefID = Spring.GetUnitDefID
 
 local updateList = {}
+local XPDefs = {}
+
+for unitDefID, unitDef in pairs(UnitDefs) do
+	if unitDef.customParams ~= nil then
+		if unitDef.customParams.rangexpscale ~= nil then
+			XPDefs[unitDefID] = unitDef
+		end
+	end
+end
 
 function gadget:GameFrame(n)
-	for unitID in pairs(updateList) do
+	for unitID, unitDef in pairs(updateList) do
```

Comment:
Use `unitDefID` as the value in the cache instead of `unitDef` as per recommendations below.

Reply:
Have done.